core/src/main/java/io/grpc/CallOptions.java

```diff
@@ -209,6 +213,75 @@ public CallOptions withExecutor(Executor executor) {
     newOptions.executor = executor;
     return newOptions;
   }
+  
+  public static final class Key<T> {
+    private final String name;
+    private final T defaultValue;
+
+    private Key(String name, T defaultValue) {
+      this.name = name;
+      this.defaultValue = defaultValue;
+    }
+
+    public T getDefault() {
+      return defaultValue;
+    }
+
+    @Override
+    public String toString() {
+      return name;
+    }
+
+    /**
+     * Factory method for creating instances of {@link Key}.
+     *
+     * @param name the name of Key.
+     * @param defaultValue default value to return when value for key not set
+     * @param <T> Key type
+     * @return Key object
+     */
+    public static <T> Key<T> of(String name, T defaultValue) {
+      return new Key<T>(name, defaultValue);
+    }
+  }
+
+  /**
+   * Set a custom option.  Any existing value for the key overridden.
+   * 
+   * @param key The option key
+   * @param value The option value.
+   */
+  public <T> CallOptions withOption(Key<T> key, T value) {
+    CallOptions newOptions = new CallOptions(this);
+    if (customOptions != null) {
```

Comment:
`customOptions` should never be `null`, right? Because it is initialized to `EMPTY_CUSTOM_OPTIONS`.